File: DropDownList/ServerSideDropDownList/ServerSideDropDownListVB/ServerSideDropDownListVB.Core/ServerAutoCompleteAppendHelper.vb
Imports Telerik.WinControls.UI
Imports System.Windows.Forms

Public Class ServerAutoCompleteAppendHelper(Of T)
    Inherits AutoCompleteAppendHelper
    Public Property Data() As IQueryable(Of T)
        Get
            Return m_Data
        End Get
        Private Set(value As IQueryable(Of T))
            m_Data = value
        End Set
    End Property
    Private m_Data As IQueryable(Of T)

    Public Sub New(owner As RadDropDownListElement, data As IEnumerable(Of T))
        MyBase.New(owner)
        Me.Data = data.AsQueryable()
        ExpressionBuilder.Instance.Optimize(Me.Data)
    End Sub

    Public Overrides Sub AutoComplete(e As KeyPressEventArgs)
        Dim findString As String = Me.CreateFindString(e)

        Dim whereExp = ExpressionBuilder.Instance.BuildStartsWithExpression(Of T)(Me.Owner.AutoCompleteValueMember, findString)
        Dim selectExp = ExpressionBuilder.Instance.BuildSelectExpression(Of T)(Me.Owner.AutoCompleteValueMember)

        Dim result As String = Me.Data.Where(whereExp).[Select](selectExp).OrderBy(Function(x) x.Length).FirstOrDefault()
        If result IsNot Nothing Then
            Owner.EditableElementText = result
            Owner.SelectionStart = findString.Length
            Owner.SelectionLength = Owner.EditableElementText.Length
            e.Handled = True
        End If
    End Sub

    Private Function CreateFindString(e As KeyPressEventArgs) As String
        Dim findString As String = ""
        If Owner.SelectionLength = 0 Then
            findString = Owner.EditableElementText + e.KeyChar
        Else
            findString = Owner.EditableElementText.Substring(0, Owner.SelectionStart) + e.KeyChar
        End If

        Return findString
    End Function
End Class

File: DropDownList/ServerSideDropDownList/ServerSideDropDownListCSharp/ServerSideDropDownList.Core/ServerAutoCompleteAppendHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Windows.Forms;
using Telerik.WinControls.UI;

namespace ServerSideDropDownList.Core
{
    public class ServerAutoCompleteAppendHelper<T> : AutoCompleteAppendHelper
    {
        public IQueryable<T> Data { get; private set; }

        public ServerAutoCompleteAppendHelper(RadDropDownListElement owner, IEnumerable<T> data)
            : base(owner)
        {
            this.Data = data.AsQueryable();
            ExpressionBuilder.Instance.Optimize(this.Data);
        }

        public override void AutoComplete(KeyPressEventArgs e)
        {
            string findString = this.CreateFindString(e);

            var whereExp = ExpressionBuilder.Instance.BuildStartsWithExpression<T>(this.Owner.AutoCompleteValueMember, findString);
            var selectExp = ExpressionBuilder.Instance.BuildSelectExpression<T>(this.Owner.AutoCompleteValueMember);

            string result = this.Data.Where(whereExp).Select(selectExp).OrderBy(x => x.Length).FirstOrDefault();
            if (result != null)
            {
                Owner.EditableElementText = result;
                Owner.SelectionStart = findString.Length;
                Owner.SelectionLength = Owner.EditableElementText.Length;
                e.Handled = true;
            }
        }

        private string CreateFindString(KeyPressEventArgs e)
        {
            string findString = "";
            if (Owner.SelectionLength == 0)
            {
                findString = Owner.EditableElementText + e.KeyChar;
            }
            else
            {
                findString = Owner.EditableElementText.Substring(0, Owner.SelectionStart) + e.KeyChar;
            }

            return findString;
        }
    }
}